core/src/main/java/org/projectnessie/cel/extension/StringsLib.java

```diff
@@ -0,0 +1,340 @@
+            Decls.newFunction(
+                INDEX_OF,
+                Decls.newInstanceOverload(
+                    "string_index_of_string",
+                    Arrays.asList(Decls.String, Decls.String),
+                    Decls.Int)),
+            Decls.newFunction(
+                LAST_INDEX_OF,
+                Decls.newInstanceOverload(
+                    "string_last_index_of_string",
+                    Arrays.asList(Decls.String, Decls.String),
+                    Decls.Int)),
+            Decls.newFunction(
+                LOWER_ASCII,
+                Decls.newInstanceOverload(
+                    "string_lower_ascii", Arrays.asList(Decls.String), Decls.String)),
+            Decls.newFunction(
+                REPLACE,
+                Decls.newInstanceOverload(
+                    "string_replace_string_string",
+                    Arrays.asList(Decls.String, Decls.String, Decls.String),
+                    Decls.String)),
+            Decls.newFunction(
+                SPLIT,
+                Decls.newInstanceOverload(
+                    "string_split_string", Arrays.asList(Decls.String, Decls.String), Decls.Dyn)),
+            Decls.newFunction(
+                SUBSTR,
+                Decls.newInstanceOverload(
+                    "string_substring_int", Arrays.asList(Decls.String, Decls.Int), Decls.String)),
+            Decls.newFunction(
+                TRIM_SPACE,
+                Decls.newInstanceOverload(
+                    "string_trim", Arrays.asList(Decls.String), Decls.String)),
+            Decls.newFunction(
+                UPPER_ASCII,
+                Decls.newInstanceOverload(
+                    "string_upper_ascii", Arrays.asList(Decls.String), Decls.String)));
+    list.add(option);
+    return list;
+  }
+
+  @Override
+  public List<ProgramOption> getProgramOptions() {
+    List<ProgramOption> list = new ArrayList<>();
+    ProgramOption functions =
+        ProgramOption.functions(
+            Overload.binary(CHAR_AT, Guards.callInStrIntOutStr(StringsLib::charAt)),
+            Overload.binary(INDEX_OF, Guards.callInStrStrOutInt(StringsLib::indexOf)),
+            Overload.binary(LAST_INDEX_OF, Guards.callInStrStrOutInt(StringsLib::lastIndexOf)),
+            Overload.unary(LOWER_ASCII, Guards.callInStrOutStr(StringsLib::lowerASCII)),
+            Overload.function(REPLACE, Guards.callInStrStrStrOutStr(StringsLib::replace)),
+            Overload.binary(SPLIT, Guards.callInStrStrOutStrArr(StringsLib::split)),
+            Overload.binary(SUBSTR, Guards.callInStrIntOutStr(StringsLib::substr)),
+            Overload.unary(TRIM_SPACE, Guards.callInStrOutStr(StringsLib::trimSpace)),
+            Overload.unary(UPPER_ASCII, Guards.callInStrOutStr(StringsLib::upperASCII)));
+    list.add(functions);
+    return list;
+  }
+
+  static String charAt(String str, int index) {
+    if (str.length() == index) {
+      return "";
+    }
+    return String.valueOf(str.charAt(index));
```

Comment:
Add `try-catch(IndexOutOfBoundsException)` and throw with a non-Java-version dependent message

Reply:
I wonder if we should wrap the Java exception with cel specific message (actually to be same as Go)

For the test cases, another option will be checking the returned value is `Err` type and cause exception is in same class. How do you think?

```
+      assertThat(((Err) result.getVal()).getCause().getClass().getCanonicalName()).isEqualTo(testData.getExceptionClass());
-      assertThat(result.getVal()).extracting(Val::toString).isEqualTo(testData.getErr());
```

Reply:
> I wonder if we should wrap the Java exception with cel specific message (actually to be same as Go)

+1